# Assignment 2

As before, if a question can be answered with 'yes/no', or a numeric value, you may simply state as much. If you incorporate code from the internet (which is not required and generally not advisable), please cite the source within your code (providing a URL is sufficient).

We will go through comparable code and concepts in the live learning sessions. If you run into trouble, start by using the help `help()` function in Python, to get information about the datasets and function in question. The internet is also a great resource when coding (though note that no outside searches are required by the assignment!). If you do incorporate code from the internet, please cite the source within your code (providing a URL is sufficient).

Please bring questions that you cannot work out on your own to office hours, work periods or share with your peers on Slack. We will work with you through the issue.

If you like, you may collaborate with others in the cohort. If you choose to do so, please indicate with whom you have worked with in your pull request by tagging their GitHub username. Separate submissions are required.

In [96]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm

# Import specific objects
from sklearn.preprocessing import StandardScaler
from ISLP import load_data

### Question 1: Classification using KNN

We'll now use the `Caravan` dataset from the `ISLP` package. (You may use `Caravan.describe()` to review details of the dataset.) In this dataset, the response variable of interest is `Purchase`, which indicates if a given customer purchased a caravan insurance policy. We will simultaneously use all other variables in the dataset to predict the response variable.

In [97]:
# Load the "Caravan" dataset using the "load_data" function from the ISLP package
Caravan = load_data('Caravan')
pd.set_option('display.max_columns', None)
Caravan.describe(include = 'all')

MOSTYPE     MAANTHUI      MGEMOMV     MGEMLEEF     MOSHOOFD  \
count   5822.000000  5822.000000  5822.000000  5822.000000  5822.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean      24.253349     1.110615     2.678805     2.991240     5.773617   
std       12.846706     0.405842     0.789835     0.814589     2.856760   
min        1.000000     1.000000     1.000000     1.000000     1.000000   
25%       10.000000     1.000000     2.000000     2.000000     3.000000   
50%       30.000000     1.000000     3.000000     3.000000     7.000000   
75%       35.000000     1.000000     3.000000     3.000000     8.000000   
max       41.000000    10.000000     5.000000     6.000000    10.000000   

             MGODRK       MGODPR       MGODOV       MGODGE       MRELGE  \
count   5822.000000  5822.000000  5822.000000  5822.000000  5822.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       0.696496     4.626932     1.069907     3.258502     6.183442   
std        1.003234     1.715843     1.017503     1.597647     1.909482   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%        0.000000     4.000000     0.000000     2.000000     5.000000   
50%        0.000000     5.000000     1.000000     3.000000     6.000000   
75%        1.000000     6.000000     2.000000     4.000000     7.000000   
max        9.000000     9.000000     5.000000     9.000000     9.000000   

             MRELSA       MRELOV     MFALLEEN     MFGEKIND     MFWEKIND  \
count   5822.000000  5822.000000  5822.000000  5822.000000  5822.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       0.883545     2.290450     1.887667     3.230333     4.300240   
std        0.965924     1.722645     1.799928     1.619973     2.005283   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%        0.000000     1.000000     0.000000     2.000000     3.000000   
50%        1.000000     2.000000     2.000000     3.000000     4.000000   
75%        1.000000     3.000000     3.000000     4.000000     6.000000   
max        7.000000     9.000000     9.000000     9.000000     9.000000   

           MOPLHOOG     MOPLMIDD     MOPLLAAG     MBERHOOG     MBERZELF  \
count   5822.000000  5822.000000  5822.000000  5822.000000  5822.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       1.461010     3.351254     4.572484     1.895053     0.397973   
std        1.622967     1.761052     2.298125     1.798321     0.775073   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%        0.000000     2.000000     3.000000     0.000000     0.000000   
50%        1.000000     3.000000     5.000000     2.000000     0.000000   
75%        2.000000     4.000000     6.000000     3.000000     1.000000   
max        9.000000     9.000000     9.000000     9.000000     5.000000   

           MBERBOER     MBERMIDD     MBERARBG     MBERARBO         MSKA  \
count   5822.000000  5822.000000  5822.000000  5822.000000  5822.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       0.522329     2.899004 

In [98]:
Caravan.shape

(5822, 86)

In [99]:
Caravan['Purchase'].values

array(['No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'No', 'No',
       'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Yes',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'No', 'No', 'No', 'No', 'No', 'No', 'Y

Before fitting any model, it is essential to understand our data. Answer the following questions about the `Caravan` dataset (Hint: use `print` and `describe`):  
_(i)_ How many observations (rows) does the dataset contain?    

5822

_(ii)_ How many variables (columns) does the dataset contain?    

86

_(iii)_ What 'variable' type is the response variable `Purchase` (e.g., 'character', 'factor', 'numeric', etc)? What are the 'levels' of the variable?  

It's an "object" in pandas dataframe. The two levels are "Yes" and "No." 

_(iv)_ How many predictor variables do we have (Hint: all variables other than `Purchase`)?  

86 - 1 = 85

Next, we must preform 'pre-processing' or 'data munging', to prepare our data for classification/prediction. For KNN, there are three essential steps. A first essential step is to 'standardize' the predictor variables. We can achieve this using the `scaler` method, provided as follows:

In [100]:
# Select predictors (excluding the 86th column)
predictors = Caravan.iloc[:, :-1]

# Standardize the predictors
scaler = StandardScaler()
predictors_standardized = pd.DataFrame(scaler.fit_transform(predictors), columns=predictors.columns)

# Display the head of the standardized predictors
print(predictors_standardized.head())

    MOSTYPE  MAANTHUI   MGEMOMV  MGEMLEEF  MOSHOOFD    MGODRK    MGODPR  \
0  0.680906  -0.27258  0.406697 -1.216964  0.779405 -0.694311  0.217444   
1  0.992297  -0.27258 -0.859500 -1.216964  0.779405  0.302552 -0.365410   
2  0.992297  -0.27258 -0.859500 -1.216964  0.779405 -0.694311 -0.365410   
3 -1.187437  -0.27258  0.406697  0.010755 -0.970980  1.299414 -0.948264   
4  1.225840  -0.27258  1.672893 -1.216964  1.479559  0.302552 -0.365410   

     MGODOV    MGODGE    MRELGE    MRELSA    MRELOV  MFALLEEN  MFGEKIND  \
0 -0.068711 -0.161816  0.427670 -0.914794 -0.168621 -0.493211 -0.759543   
1 -0.068711  0.464159 -0.096077  1.155941 -0.168621 -1.048836  0.475152   
2  0.914172  0.464159 -1.667319  1.155941  0.992484  1.173666  0.475152   
3  0.914172  0.464159 -0.619824  1.155941 -0.168621  0.062415 -0.142196   
4 -0.068711  0.464159  0.427670  0.120574 -0.168621  0.062415  0.475152   

   MFWEKIND  MOPLHOOG  MOPLMIDD  MOPLLAAG  MBERHOOG  MBERZELF  MBERBOER  \
0  0.847713 -0.284078 -

_(v)_ Why is it important to standardize the predictor variables?  

The KNN method classifies observations by finding the classifications of their nearest neighbours, with an assumption that the distances between observations are measured by the same unit. Therefore, it is important that the distances between variables are "standardized" by the same scale.

_(vi)_ Why did we elect not to standard our response variable `Purchase`?  

Each explanatory variable bears a potentially predictive relationship with the same response variable 'Purchase.' Because the same variable is being compared, it does not need to be standardized.


_(vii)_ A second essential step is to set a random seed. Do so below (Hint: use the `random.seed` function). Why is setting a seed important? Is the particular seed value important? Why or why not?

The random.seed function ensures that the random module creates the same set of 'random' numbers every time. A particular seed value is not necessary, but a consistent value is.

In [101]:
# please see below

In [102]:
# Create a random vector of True and False values
np.random.seed(1893)
split = np.random.choice([True, False], size=len(predictors_standardized), replace=True, p=[0.75, 0.25])
import sys
np.set_printoptions(threshold=sys.maxsize)
split

array([ True,  True,  True,  True, False,  True, False, False,  True,
        True, False, False,  True,  True, False,  True,  True,  True,
       False,  True,  True, False,  True, False,  True, False,  True,
        True,  True, False,  True, False,  True,  True,  True, False,
       False,  True,  True, False,  True,  True, False, False,  True,
        True,  True,  True, False,  True, False,  True,  True,  True,
        True, False,  True,  True,  True,  True, False,  True,  True,
        True,  True, False, False,  True,  True,  True, False,  True,
       False,  True,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True, False,  True,
       False,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True, False,  True,  True,  True,  True,  True,  True,  True,
       False,  True,

_(viii)_ A third essential step is to split our standardized data into separate training and testing sets. We will split into 75% training and 25% testing. The provided code randomly partitions our data, and creates linked training sets for the predictors and response variables. Extend the code to create a non-overlapping test set for the predictors and response variables.

In [103]:
# Create a random vector of True and False values
#split = np.random.choice([True, False], size=len(predictors_standardized), replace=True, p=[0.75, 0.25])

# Define the training set for X (predictors)
training_X = predictors_standardized[split]

# Define the training set for Y (response)
training_Y = Caravan.loc[split, 'Purchase']

# Define the testing set for X (predictors)
testing_X = predictors_standardized[~split]

# Define the testing set for Y (response)
testing_Y = Caravan.loc[~split, 'Purchase']


_(ix)_ We are finally set to fit the KNN model. In Python, we can use the `KNeighborsClassifier()` function. Fit the KNN with k=1. (You may review arguments to knn by typing `help(knn.fit)`). 

In [113]:
from sklearn.neighbors import KNeighborsClassifier
knn1 = KNeighborsClassifier(n_neighbors=1)
training_X, testing_X = [np.asarray(X) for X in [training_X, testing_X]]
knn1.fit(training_X, training_Y)
knn1_pred = knn1.predict(testing_X)


Using your fit model, answer the following questions:   
_(x)_ What is the prediction accuracy? (Hint: use the `score` method, and compare your model to `testing_Y`)  

88.9%

_(xi)_ What is the predictor error ? (Hint: compute it from the accuracy)

100% - 89.5% = 10.5%

In [114]:
knn1.score(testing_X, testing_Y)

0.895542248835662

In [115]:
1 - knn1.score(testing_X, testing_Y)

0.10445775116433798

_(xii)_ How does this prediction error/accuracy compare to what could be achieved via random guesses? To answer this, consider the percent of customers in the `Caravan` dataset who actually purchase insurance, computed below:

Random guessing would predict a purchase rate of ~50%. Considering that the actual purchase rate is ~6.4%, and the model is ~89% accurate (i.e. predicting correctly in ~6.4%+/-10% of cases), it is far from being random.

In [107]:
# Calculate the percentage of customers who purchase insurance
percentage_purchase = (Caravan['Purchase'].eq('Yes').sum() / len(Caravan['Purchase'])) * 100

print(percentage_purchase)

6.357325538911216


_(xiii)_ Fit a second KNN model, with $K=3$. Does this model perform better (i.e., have higher accuracy, compared to a random guess)?

Yes, the accuracy is increased from ~89.5% to ~93.1%.

In [116]:
knn2 = KNeighborsClassifier(n_neighbors=3)
training_X, testing_X = [np.asarray(X) for X in [training_X, testing_X]]
knn2.fit(training_X, training_Y)
knn2_pred = knn2.predict(testing_X)
knn2.score(testing_X, testing_Y)

0.9314703925482368

# Criteria

|Criteria            |Complete           |Incomplete          |
|--------------------|---------------|--------------|
|Classification using KNN|All steps are done correctly and the answers are correct.|At least one step is done incorrectly leading to a wrong answer.|

## Submission Information

🚨 **Please review our [Assignment Submission Guide](https://github.com/UofT-DSI/onboarding/blob/main/onboarding_documents/submissions.md)** 🚨 for detailed instructions on how to format, branch, and submit your work. Following these guidelines is crucial for your submissions to be evaluated correctly.

### Submission Parameters:
* Submission Due Date: `HH:MM AM/PM - DD/MM/YYYY`
* The branch name for your repo should be: `assignment-2`
* What to submit for this assignment:
    * This Jupyter Notebook (assignment_2.ipynb) should be populated and should be the only change in your pull request.
* What the pull request link should look like for this assignment: `https://github.com/<your_github_username>/applying_statistical_concepts/pull/<pr_id>`
    * Open a private window in your browser. Copy and paste the link to your pull request into the address bar. Make sure you can see your pull request properly. This helps the technical facilitator and learning support staff review your submission easily.

Checklist:
- [ ] Created a branch with the correct naming convention.
- [ ] Ensured that the repository is public.
- [ ] Reviewed the PR description guidelines and adhered to them.
- [ ] Verify that the link is accessible in a private browser window.

If you encounter any difficulties or have questions, please don't hesitate to reach out to our team via our Slack at `#cohort-3-help`. Our Technical Facilitators and Learning Support staff are here to help you navigate any challenges.
